src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs

```diff
@@ -557,37 +658,34 @@ protected virtual void ValidateNonce(OpenIdConnectProtocolValidationContext vali
         /// </summary>
         /// <param name="validationContext">a <see cref="OpenIdConnectProtocolValidationContext"/> that contains the 'state' to validate.</param>
         /// <exception cref="ArgumentNullException">if 'validationContext' is null.</exception>
+        /// <exception cref="OpenIdConnectProtocolInvalidStateException">if 'state' is present in <see cref="OpenIdConnectProtocolValidationContext.State"/> but either <see cref="OpenIdConnectProtocolValidationContext.ProtocolMessage"/> or <see cref="OpenIdConnectProtocolValidationContext.ProtocolMessage.State"/> is null.</exception>
         /// <exception cref="OpenIdConnectProtocolInvalidStateException">if 'state' in the context does not match the state in the message.</exception>
         protected virtual void ValidateState(OpenIdConnectProtocolValidationContext validationContext)
         {
-
             if (validationContext == null)
             {
                 LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10000, GetType() + ": validationContext"), typeof(ArgumentNullException), EventLevel.Verbose);
             }
 
             // if state is missing, but not required just return. Otherwise process it.
-            if (string.IsNullOrEmpty(validationContext.State))
+            if (RequireStateValidation && string.IsNullOrEmpty(validationContext.State))
```

Comment:
Same comments as above with nonce, if context has state but not message, that is an error.